### DISCUSSION TITLE
How to determine push/pop/replace on router event callbacks? (re-open #3077)

### DISCUSSION THREAD

**Discussion by nol13** [vercel/next.js#19217] [Help]
Was asked here without an answer ever really being given. 
https://github.com/vercel/next.js/issues/3077

Is there a reason why the router events pass only the url and can't also pass the PUSH/POP/REPLACE action being done, like is the case with react-router? Is there some super obvious way to do this that I'm missing without overriding the router with custom functions or some complex counter tracking history.length changes to differentiate push from replace? 

I guess wondering if there is a technical reason that next can't have..?

 `router.events.on('routeChangeComplete', (url, actionType) => {/* do stuff */})`

Edit: I guess even react-router won't tell you if it's a forward POP or a backward POP on it's own, but still curious to see if anyone has been able to implement this or had a need for it.

**Comment by agusterodin** [comment]
Did you ever find a solution?

My use case is a modal page where I want a close button to appear at the top if it is the first page loaded. Clicking close will peform push("/") to take the user to the homepage.

If the page was visited from a link within the application, I want a back button to appear at the top instead which will perform a history.back().

Basically I would want to keep a value in state called pageNavigationDepth which gets incremented by 1 when push is used and gets decremented by 1 if pop is used. The button at the top of the page (close or back) will be dictated by whether the pageNavigationDepth value is 0.

**Comment by agusterodin** [comment]
The router gives us a beforePopState callback function. If only there was also a beforePushState function.
